### PR TITLE
Emit correct event user when deleting user

### PR DESCRIPTION
**Pull request by kimlisa**
#### Description
In order to capture the correct event user when deleting users, it is required to pass around contexts:
- created DeleteUser for grpc api
- updated all DeleteUser uses signature to accept context
- update unit tests

#### Screenshot
Before changes:
Screenshot from 2020-05-13 09-23-57

After changes, captures the correct user:
screenshot

### CODE REVIEW

**Review by fspmarshall — COMMENTED** [review]
GRPC method implementation looks good, but you need to take cross-version compatibility into account.  We need at least a one-version grace-period where clients can talk to old servers and servers can serve old clients.

Instead of deleting the HTTP logic, keep it around with a `DELETE IN: 5.1` (or similar) directive.  The server can simply expose HTTP and GRPC methods simultaneously.  The client should attempt a GRPC call first.  If it gets back an `UNIMPLEMENTED` error, it should fallback to the old HTTP logic.

You can find an example of compatibility fallback logic [here](https://github.com/gravitational/teleport/blob/e59afe591a371a81f3663681a12f08cec9b5cf83/lib/auth/clt.go#L1474-L1497).

**Review by alex-kovoy — COMMENTED** [review]
Why we are replacing existing HTTP method as part of this ticket? Lets add audit events first. Plus, both methods (http, grpc) should exist side by side for 2next releases, so lets focus on events first.

File: lib/auth/apiserver.go [hunk 1]
```diff
@@ -896,12 +896,13 @@ func (s *APIServer) getUsers(auth ClientI, w http.ResponseWriter, r *http.Reques
 	return out, nil
 }
 
+// DELETE IN: 5.2 REST method is replaced by grpc method with context.
 func (s *APIServer) deleteUser(auth ClientI, w http.ResponseWriter, r *http.Request, p httprouter.Params, version string) (interface{}, error) {
 	user := p.ByName("user")
-	if err := auth.DeleteUser(user); err != nil {
+	if err := auth.DeleteUser(context.TODO(), user); err != nil {
```
Comment (awly) [inline lib/auth/apiserver.go:902]: use `r.Context()` here instead

File: lib/auth/clt.go [hunk 2]
```diff
@@ -1567,10 +1567,31 @@ func (c *Client) grpcGetUsers(withSecrets bool) ([]services.User, error) {
...
-	return trace.Wrap(err)
+// DeleteUser deletes a user by username.
+func (c *Client) DeleteUser(ctx context.Context, user string) error {
+	clt, err := c.grpc()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	req := &proto.DeleteUserRequest{Name: user}
+	_, grpcErr := clt.DeleteUser(ctx, req)
```
Comment (awly) [inline lib/auth/clt.go:1578]: change `grpcErr` to `err` unless it needs to be different

**Review by awly — COMMENTED** [review]
Please add a unit test for backwards-compatibility logic

File: lib/auth/apiserver.go (hunk 1 above)
Comment (kimlisa) [reply lib/auth/apiserver.go:902]: revised

File: lib/auth/clt.go (hunk 2 above)
Comment (kimlisa) [reply lib/auth/clt.go:1578]: revised

File: lib/auth/clt.go [hunk 3]
```diff
@@ -70,6 +70,21 @@ const (
 // Dialer defines dialer function
 type Dialer func(network, addr string) (net.Conn, error)
 
+// IClientServices defines wrapper methods for roundtrip package.
+// Methods that are not wrappers are dependent on the roundtrip.Client instance
+// held by ClientService struct (which implements this interface).
+//
+// Interface was made for testing backward compatability logic between GRPC and REST.
```
Comment (awly) [inline lib/auth/clt.go:77]: `s/compatability/compatibility`

File: lib/auth/clt.go [hunk 4]
```diff
@@ -70,6 +70,21 @@ const (
...
+type IClientServices interface {
```
Comment (awly) [inline lib/auth/clt.go:78]: Don't prefix type names with the kind of type (like `I` here). Also, this type doesn't need to be exported.
How about `type rpcClient`? Then `ClientServices` below can be renamed to `restRPCClient`.

File: lib/auth/clt.go [hunk 5]
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
 	return *rsp, nil
 }
 
+// ClientServices implements IClientServices interface.
+type ClientServices struct {
+	roundtrip.Client
+}
+
+// PostJSON is a generic method that issues http POST request to the server
+func (c *ClientServices) PostJSON(
```
Comment (awly) [inline lib/auth/clt.go:2659]: don't wrap this line https://github.com/golang/go/wiki/CodeReviewComments#line-length
similarly for other methods below

File: lib/auth/clt.go [hunk 6]
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
...
+
+// Delete issues http Delete Request to the server
+func (c *ClientServices) Delete(u string) (*roundtrip.Response, error) {
+	return httplib.ConvertResponse(c.Client.Delete(context.TODO(), u))
+}
+
+// Endpoint returns a URL constructed from parts and version appended, e.g.
+func (c *ClientServices) Endpoint(params ...string) string {
+	return c.Client.Endpoint(params...)
+}
```
Comment (awly) [inline lib/auth/clt.go:2691]: you can skip this, since `c.Client` is embedded

File: lib/auth/clt.go [hunk 7]
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
...
+
+// PostSessionSlice allows clients to submit session stream chunks to the audit log
+// (part of evets.IAuditLog interface)
+//
+// The data is POSTed to HTTP server as a simple binary body (no encodings of any
+// kind are needed)
+func (c *ClientServices) PostSessionSlice(slice events.SessionSlice) error {
```
Comment (awly) [inline lib/auth/clt.go:2698]: this method seems out-of-place here.
all others are generic while this is type-specific.

how about a new method `PostBytes` and another method `func (c *Client) PostSessionSlice` that calls into this?

File: lib/auth/clt.go [hunk 8]
```diff
@@ -2670,6 +2650,73 @@ func (c *Client) Ping(ctx context.Context) (proto.PingResponse, error) {
...
+	data, err := slice.Marshal()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	r, err := http.NewRequest("POST", c.Endpoint("namespaces", slice.Namespace, "sessions", slice.SessionID, "slice"), bytes.NewReader(data))
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	r.Header.Set("Content-Type", "application/grpc")
```
Comment (awly) [inline lib/auth/clt.go:2707]: Hmm, why is this `application/grpc`?
If it's actually gRPC, should it go over the gRPC transport?

File: lib/auth/mock.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	empty "github.com/golang/protobuf/ptypes/empty"
+	"github.com/gravitational/roundtrip"
+	"github.com/gravitational/teleport/lib/auth/proto"
+	"github.com/gravitational/trace"
+	"github.com/gravitational/trace/trail"
+	"google.golang.org/grpc"
+)
+
+// MockClient stubs the IClientServices interface.
+type MockClient struct {
```
Comment (awly) [inline lib/auth/mock.go:30]: will this get used outside of `lib/auth`?
if not, it should probably live in a `_test.go` file and be unexported.

File: lib/auth/mock.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	return nil, nil
+}
+
+// Reset resets states to zero value.
+func (m *MockClient) Reset() {
+	m.MethodCalled = false
+}
+
+// MockAuthServiceClient stubs grpc AuthServiceClient interface.
+type MockAuthServiceClient struct {
```
Comment (awly) [inline lib/auth/mock.go:47]: same here, it should live in a `_test.go` file and be unexported.

File: lib/auth/mock.go [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+	proto.AuthServiceClient
+}
+
+// DeleteUser is a mock that returns an error.
+func (m *MockAuthServiceClient) DeleteUser(ctx context.Context, in *proto.DeleteUserRequest, opts ...grpc.CallOption) (*empty.Empty, error) {
+	return nil, trail.ToGRPC(trace.NotImplemented(""))
```
Comment (awly) [inline lib/auth/mock.go:53]: to make this more flexible, I suggest making the returned error a field of the struct.
that way you can experiment with other kinds of errors.

File: lib/auth/tls_test.go [hunk 12]
```diff
@@ -2560,3 +2560,16 @@ func (s *TLSSuite) TestEventsClusterConfig(c *check.C) {
 	c.Assert(err, check.IsNil)
 	suite.ExpectResource(c, w, 3*time.Second, clusterNameResource)
 }
+
+func (s *TLSSuite) TestAPIBackwardsCompatibilityLogic(c *check.C) {
```
Comment (awly) [inline lib/auth/tls_test.go:2564]: add a case where gRPC server returns a different kind of error (not `NotImplemented`)

File: lib/auth/clt.go (hunk 3 above)
Comment (kimlisa) [reply lib/auth/clt.go:77]: revised

File: lib/auth/clt.go (hunk 5 above)
Comment (kimlisa) [reply lib/auth/clt.go:2659]: revised

File: lib/auth/clt.go (hunk 4 above)
Comment (kimlisa) [reply lib/auth/clt.go:78]: revised, but had a question, is rest a type of rpc? I searched the internet and they seem to distinguish between the two

File: lib/auth/mock.go (hunk 10 above)
Comment (kimlisa) [reply lib/auth/mock.go:47]: revised (i placed in top of `tls_test.go`)

File: lib/auth/mock.go (hunk 11 above)
Comment (kimlisa) [reply lib/auth/mock.go:53]: revised

File: lib/auth/tls_test.go (hunk 12 above)
Comment (kimlisa) [reply lib/auth/tls_test.go:2564]: revised

File: lib/auth/clt.go (hunk 4 above)
Comment (awly) [reply lib/auth/clt.go]: technically, REST is more of an API design concept.
but in practice it almost always means: "RPC over HTTP with URLs and HTTP verbs identifying the kind of operation to be performed".
so I tend to think of REST as a kind of RPC transport layer.
an alternative name here would be `httpRPCClient`

File: lib/auth/clt.go (hunk 8 above)
Comment (alex-kovoy) [reply lib/auth/clt.go]: @klizhentas do you recall why you had to explicitly set these headers?

File: lib/auth/tls_test.go [hunk 13]
```diff
@@ -45,13 +47,35 @@ import (
...
 	"github.com/gravitational/trace"
 	"github.com/jonboulle/clockwork"
 	"github.com/pquerna/otp/totp"
 	"gopkg.in/check.v1"
 )
 
+// MockAuthServiceClient stubs grpc AuthServiceClient interface.
+type mockAuthServiceClient struct {
+	proto.AuthServiceClient
+	errType error
```
Comment (awly) [inline lib/auth/tls_test.go:61]: nit: how about `err` or `deleteUserErr`?

**Review by awly — APPROVED** [review]

File: lib/auth/tls_test.go (hunk 13 above)
Comment (kimlisa) [reply lib/auth/tls_test.go:61]: revised to `err`

**Review by russjones — APPROVED** [review]

**Pull request merged**